### PR TITLE
Feature/migl/rdphoen 1140 adapt new volumes for swupdate

### PR DESCRIPTION
**Pull request by linkjumper**
This PR makes the swu images work again with the new partitioning / volumes.

- tested for imx8mpevk with maintenance image
- Jenkins IRMA6-Release 1: not needed

## Build

```
# make sure meta-swupdate points to at least 83f5b590476130cd8c3ba8984119692d7553af87
$ bitbake mc:imx8mp-evk:irma6-maintenance-uuu
$ bitbake mc:imx8mp-evk:irma6-maintenance-swuimage
```


## Prepare

Flash image with uuu:
```
# set jumper to serial download mode
$ cd build/tmp/deploy/images/imx8mpevk/irma6-maintenance-uuu
$ uuu flashall.uuu
# unset jumper to serial download mode
```

Copy image onto sensor with netcat
```
# Target:
$ cd /mnt/iris/
$ nc -l -p 1234 > out.file

# Host
$ nc -w 3 <SENSOR_IP> 1234 < ../irma6-maintenance-swuimage-imx8mpevk.swu
# Wait 3..5s [CTRL+C] to stop sending
```

## Test

**1) Check if u-boot-env's are set/unset before/after each update test**

```
# upgrade_available: Bootcounter write enable flag
#   0 or unset : bootcounter disabled
#   1: bootcounter enabled
# ustate: Bootloader state
#          0 or unset : No update pending
#          1          : Update installed
#          2          : Under Testing
#          3          : Failed

# Before
root@imx8mpevk:/mnt/iris# fw_printenv ustate
ustate=                                                  <-- unset == no update pending (only on fresh flashed images otherwise unknown)
root@imx8mpevk:/mnt/iris# fw_printenv upgrade_available
upgrade_available=                                       <-- unset == no upgrade available

# After
root@imx8mpevk:/mnt/iris# fw_printenv ustate
ustate=1                                                  <-- 1 == update installed
root@imx8mpevk:/mnt/iris# fw_printenv upgrade_available
upgrade_available=1                                       <-- 1 == upgrade available

# Manual unset
root@imx8mpevk:/mnt/iris# fw_setenv upgrade_available                                                                                                                                                                                                                         
root@imx8mpevk:/mnt/iris# fw_setenv ustate
```

**2) A -> B**
```
# Boot firmware A (default is A so skip)
#u-boot=> editenv mmcargs
#edit: setenv bootargs ${jh_clk} console=${console} root=${mmcroot} firmware_a
#u-boot=> boot
Starting kernel ...
[..]
Initramfs Bootstrap...
Kernel cmdline: console=ttymxc1,115200 root=/dev/mmcblk2p-22 rootwait rw firmware_a
Select firmware_a
Root device: /dev/mapper/irma6lvm-rootfs_a
Verity device: /dev/mapper/verity-rootfs_a
Switch to root
[..]

$ mkdir /tmp/boot
$ mkdir /tmp/key
$ mkdir /tmp/rootfs

# Remove Firmware B data before update
$ mount /dev/mmcblk2p3 /tmp/boot
$ cat /dev/urandom > /dev/mapper/irma6lvm-rootfs_b # just wait a few sec!!
$ cat /dev/urandom > /dev/mapper/irma6lvm-rootfs_b_hash
$ mount /dev/mapper/irma6lvm-keystore /tmp/key
$ rm /tmp/boot/fitImage.signed
$ rm /tmp/key/rootfs_b_roothash
$ umount /tmp/boot
$ umount /tmp/key

# Run update
$ swupdate -v -i out.file
[..]
[TRACE] : SWUPDATE running :  [read_lines_notify] : Update firmware_b
[..]
[INFO ] : SWUPDATE running :  [endupdate] : Swupdate was successful !

# Verify update is written
$ mount /dev/mmcblk2p3 /tmp/boot
$ mount /dev/mapper/irma6lvm-rootfs_b /tmp/rootfs/
$ mount /dev/mapper/irma6lvm-keystore /tmp/key
$ ls -la /tmp/boot/fitImage.signed                         <-- ok :)
$ ls -la /tmp/rootfs/                                      <-- ok :)
$ ls -la /tmp/key/rootfs_b_roothash                        <-- ok :)
$ umount /tmp/boot
$ umount /tmp/rootfs
$ umount /tmp/key

# Finally boot into B: reboot + go on with 3)
```

**3) B -> A**

```
# Boot firmware B
u-boot=> editenv mmcargs
edit: setenv bootargs ${jh_clk} console=${console} root=${mmcroot} firmware_b
u-boot=> boot
Starting kernel ...
[..]
Initramfs Bootstrap...
Kernel cmdline: console=ttymxc1,115200 root=/dev/mmcblk2p-22 rootwait rw firmware_b
Select firmware_b
Root device: /dev/mapper/irma6lvm-rootfs_b
Verity device: /dev/mapper/verity-rootfs_b
Switch to root
[..]

# Remove Firmware A before install like 3) use /dev/mmcblk2p2 (/boot) and "_a" for /dev/mapper

# Run update
$ swupdate -v -i out.file
[..]
[TRACE] : SWUPDATE running :  [read_lines_notify] : Update firmware_a  <-- !!!
[..]
[INFO ] : SWUPDATE running :  [endupdate] : Swupdate was successful !

# Verify update is written as in 2)
# Finally boot into A again: reboot + append "firmware_a" in u-boot's mmcargs
```

# Full log from A -> B as reference

```
root@imx8mpevk:/mnt/iris# swupdate -v -i out.file                                                                                                                                                                                                                             
Swupdate v2021.11.0                                                                                                                                                                                                                                                           
                                                                                                                                                                                                                                                                              
Licensed under GPLv2. See source distribution for detailed copyright notices.                                                                                                                                                                                                 
                                                                                                                                                                                                                                                                              
[INFO ] : SWUPDATE running :  [main] : Running on imx8mpevk Revision 1.0                                                                                                                                                                                                      
[INFO ] : SWUPDATE running :  [print_registered_handlers] : Registered handlers:                                                                                                                                                                                              
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     dummy                                                                                                                                                                                                         
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     uboot                                                                                                                                                                                                         
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     bootloader                                                                                                                                                                                                    
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     flash                                                                                                                                                                                                         
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     lua                                                                                                                                                                                                           
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     raw                                                                                                                                                                                                           
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     rawfile                                                                                                                                                                                                       
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     rawcopy                                                                                                                                                                                                       
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     shellscript                                                                                                                                                                                                   
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     preinstall                                                                                                                                                                                                    
[INFO ] : SWUPDATE running :  [print_registered_handlers] :     postinstall                                                                                                                                                                                                   
[TRACE] : SWUPDATE running :  [listener_create] : creating socket at /tmp/swupdateprog                                                                                                                                                                                        
[TRACE] : SWUPDATE running :  [network_initializer] : Main loop daemon                                                                                                                                                                                                        
[TRACE] : SWUPDATE running :  [listener_create] : creating socket at /tmp/sockinstctrl                                                                                                                                                                                        
[TRACE] : SWUPDATE running :  [network_thread] : Incoming network request: processing...                                                                                                                                                                                      
[INFO ] : SWUPDATE started :  Software Update started !                                                                                                                                                                                                                       
[TRACE] : SWUPDATE running :  [network_initializer] : Software update started                                                                                                                                                                                                 
[TRACE] : SWUPDATE running :  [extract_file_to_tmp] : Found file                                                                                                                                                                                                              
[TRACE] : SWUPDATE running :  [extract_file_to_tmp] :   filename sw-description                                                                                                                                                                                               
[TRACE] : SWUPDATE running :  [extract_file_to_tmp] :   size 1281                                                                                                                                                                                                             
[TRACE] : SWUPDATE running :  [get_common_fields] : Version 0.1.0                                                                                                                                                                                                             
[TRACE] : SWUPDATE running :  [get_common_fields] : Description Firmware update for IRMA6 image                                                                                                                                                                               
[TRACE] : SWUPDATE running :  [get_common_fields] : Setting bootloader transaction marker: false                                                                                                                                                                              
[TRACE] : SWUPDATE running :  [parse_hw_compatibility] : Accepted Hw Revision : 1.0                                                                                                                                                                                           
[TRACE] : SWUPDATE running :  [add_properties] : Found properties for irma6-fitimage.itb.signed:                                                                                                                                                                              
[TRACE] : SWUPDATE running :  [add_properties_cb] :             Property create-destination: true                                                                                                                                                                             
[TRACE] : SWUPDATE running :  [_parse_files] : Found File 5.4.147: irma6-fitimage.itb.signed --> fitImage.signed (/dev/swu_kernel)                                                                                                                                            
[TRACE] : SWUPDATE running :  [_parse_images] : Found Image: bootloader-env in device :  for handler bootloader                                                                                                                                                               
[TRACE] : SWUPDATE running :  [_parse_images] : Found compressed Image irma6-maintenance-2.0.5-dev-20220605110531: irma6-maintenance-imx8mpevk.ext4.gz in device : /dev/swu_rootfs for handler raw                                                                            
[TRACE] : SWUPDATE running :  [_parse_scripts] : Found Script: pre_post_inst.sh                                                                                                                                                                                               
[TRACE] : SWUPDATE running :  [check_hw_compatibility] : Hardware imx8mpevk Revision: 1.0                                                                                                                                                                                     
[TRACE] : SWUPDATE running :  [check_hw_compatibility] : Hardware compatibility verified                                                                                                                                                                                      
[TRACE] : SWUPDATE running :  [extract_files] : Found file                                                                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [extract_files] :         filename pre_post_inst.sh                                                                                                                                                                                             
[TRACE] : SWUPDATE running :  [extract_files] :         size 1953 required                                                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [extract_files] : Found file                                                                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [extract_files] :         filename bootloader-env                                                                                                                                                                                               
[TRACE] : SWUPDATE running :  [extract_files] :         size 20 required                                                                                                                                                                                                      
[TRACE] : SWUPDATE running :  [extract_files] : Found file                                                                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [extract_files] :         filename irma6-fitimage.itb.signed                                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [extract_files] :         size 39279196 required                                                                                                                                                                                                
[TRACE] : SWUPDATE running :  [extract_files] : Found file                                                                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [extract_files] :         filename irma6-maintenance-imx8mpevk.ext4.gz                                                                                                                                                                          
[TRACE] : SWUPDATE running :  [extract_files] :         size 16811303 required                                                                                                                                                                                                
[TRACE] : SWUPDATE running :  [extract_padding] : Expecting 92 padding bytes at end-of-file                                                                                                                                                                                   
[TRACE] : SWUPDATE running :  [network_initializer] : Valid image found: copying to FLASH                                                                                                                                                                                     
[INFO ] : SWUPDATE running :  Installation in progress                                                                                                                                                                                                                        
[TRACE] : SWUPDATE running :  [read_lines_notify] : Update firmware_b                                                                                                                                                                                                         
[TRACE] : SWUPDATE running :  [run_system_cmd] : /tmp/scripts/pre_post_inst.sh preinst  command returned 0                                                                                                                                                                    
[TRACE] : SWUPDATE running :  [install_single_image] : Found installer for stream irma6-maintenance-imx8mpevk.ext4.gz raw                                                                                                                                                     
[TRACE] : SWUPDATE running :  [install_single_image] : Found installer for stream bootloader-env bootloader                                                                                                                                                                   
[TRACE] : SWUPDATE running :  [install_boot_environment] : name = upgrade_available value = 1                                                                                                                                                                                 
[TRACE] : SWUPDATE running :  [install_single_image] : Found installer for stream irma6-fitimage.itb.signed rawfile                                                                                                                                                           
[TRACE] : SWUPDATE running :  [install_raw_file] : Installing file irma6-fitimage.itb.signed on /tmp/datadst/fitImage.signed                                                                                                                                                  
[TRACE] : SWUPDATE running :  [install_raw_file] : Creating path /tmp/datadst/fitImage.signed                                                                                                                                                                                 
[TRACE] : SWUPDATE running :  [read_lines_notify] : Update firmware_b                                                                                                                                                                                                         
[TRACE] : SWUPDATE running :  [run_system_cmd] : /tmp/scripts/pre_post_inst.sh postinst  command returned 0                                                                                                                                                                   
[INFO ] : SWUPDATE successful ! SWUPDATE successful !                                                                                                                                                                                                                         
[TRACE] : SWUPDATE running :  [network_initializer] : Main thread sleep again !                                                                                                                                                                                               
[INFO ] : No SWUPDATE running :  Waiting for requests...                                                                                                                                                                                                                      
[INFO ] : SWUPDATE running :  [endupdate] : Swupdate was successful !
```

### CODE REVIEW

**Comment by ThomasWNoack** [comment]
Tried Update via PA:
```
SWUPDATE: Software Update started !
SWUPDATE: [network_initializer] : Software update started
SWUPDATE: [extract_file_to_tmp] : Found file
SWUPDATE: [extract_file_to_tmp] :  filename sw-description
SWUPDATE: [extract_file_to_tmp] :  size 1281
SWUPDATE: [get_common_fields] : Version 0.1.0
SWUPDATE: [get_common_fields] : Description Firmware update for IRMA6 image
SWUPDATE: [get_common_fields] : Setting bootloader transaction marker: false
SWUPDATE: [parse_hw_compatibility] : Accepted Hw Revision : 1.0
SWUPDATE: [add_properties] : Found properties for irma6-fitimage.itb.signed:
SWUPDATE: [add_properties_cb] :   Property create-destination: true
SWUPDATE: [_parse_files] : Found File 5.4.147: irma6-fitimage.itb.signed --> fitImage.signed (/dev/swu_kernel)
SWUPDATE: [_parse_images] : Found Image: bootloader-env in device :  for handler bootloader
SWUPDATE: [_parse_images] : Found compressed Image irma6-maintenance-2.0.5-dev-20220607125459: irma6-maintenance-imx8mpevk.ext4.gz in device : /dev/swu_rootfs for handler raw
SWUPDATE: [_parse_scripts] : Found Script: pre_post_inst.sh
SWUPDATE: [check_hw_compatibility] : Hardware imx8mpevk Revision: 1.0
SWUPDATE: [check_hw_compatibility] : Hardware compatibility verified
SWUPDATE: [extract_files] : Found file
SWUPDATE: [extract_files] :  filename pre_post_inst.sh
SWUPDATE: [extract_files] :  size 2087 required
SWUPDATE: [extract_files] : Found file
SWUPDATE: [extract_files] :  filename bootloader-env
SWUPDATE: [extract_files] :  size 20 required
SWUPDATE: [extract_files] : Found file
SWUPDATE: [extract_files] :  filename irma6-fitimage.itb.signed
SWUPDATE: [extract_files] :  size 39279196 required
SWUPDATE: [extract_files] : Found file
SWUPDATE: [extract_files] :  filename irma6-maintenance-imx8mpevk.ext4.gz
SWUPDATE: [extract_files] :  size 20507646 required
SWUPDATE: [extract_padding] : Expecting 388 padding bytes at end-of-file
SWUPDATE: [network_initializer] : Valid image found: copying to FLASH
SWUPDATE: Installation in progress
SWUPDATE: ERROR : /tmp/scripts/pre_post_inst.sh: line 18: /usr/bin/grep: No such file or directory
SWUPDATE: [read_lines_notify] : Update firmware_b
SWUPDATE: [run_system_cmd] : /tmp/scripts/pre_post_inst.sh preinst  command returned 0
SWUPDATE: [install_single_image] : Found installer for stream irma6-maintenance-imx8mpevk.ext4.gz raw
SWUPDATE: [install_single_image] : Found installer for stream bootloader-env bootloader
SWUPDATE: [install_boot_environment] : name = upgrade_available value = 1
SWUPDATE: [install_single_image] : Found installer for stream irma6-fitimage.itb.signed rawfile
SWUPDATE: [install_raw_file] : Installing file irma6-fitimage.itb.signed on /tmp/datadst/fitImage.signed
SWUPDATE: [install_raw_file] : Creating path /tmp/datadst/fitImage.signed
SWUPDATE: ERROR : /tmp/scripts/pre_post_inst.sh: line 18: /usr/bin/grep: No such file or directory
SWUPDATE: [read_lines_notify] : Update firmware_b
SWUPDATE: ERROR : /tmp/scripts/pre_post_inst.sh: line 64: /usr/bin/mkdir: No such file or directory
SWUPDATE: [run_system_cmd] : /tmp/scripts/pre_post_inst.sh postinst  command returned 1
SWUPDATE: ERROR : execute postinstall scripts failed
SWUPDATE: Installation failed !
SWUPDATE: [bootloader_initialize] : Cannot read environment, using default
SWUPDATE: [network_initializer] : Main thread sleep again !
SWUPDATE: Waiting for requests...
```
uuu and swu built locally.

**Comment by ThomasWNoack** [comment]
Side observation: PA was not started automatically

File: recipes-support/irma6-swuimage/files/pre_post_inst.sh [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+cmds_exist () {
+	exists veritysetup
+	exists sfdisk
+	exists jq
+}
+
+parse_cmdline() {
+	grep -q 'linuxboot_b\|firmware_b' /proc/cmdline
+	if [ $? -eq 0 ]; then
```
Comment (iris-ersc) [inline recipes-support/irma6-swuimage/files/pre_post_inst.sh:19]: Can be combined to:

```
if grep -q 'linuxboot_b\|firmware_b' /proc/cmdline
then
```

See: https://www.shellcheck.net/wiki/SC2181

**Review by iris-ersc — CHANGES_REQUESTED** [review]
Please add package name to commit "Add sha256 check"

**Review by sven-foerster — APPROVED** [review]

**Review by iris-ersc — APPROVED** [review]

**Comment by iris-ersc** [comment]
## Maintainer Checklist
- Is there a changelog entry? *Yes*
- Do the commit messages comply with the specifications? *Yes*
- Was a Jenkins run performed and was it error-free? *Yes*
- Is a manual rebase necessary? *already done*

**Pull request merged**